isisd/isis_flex_algo.c

```diff
@@ -327,4 +329,162 @@ bool isis_flex_algo_constraint_drop(struct isis_spftree *spftree,
 	return false;
 }
 
+/*
+ * Flex-Algo Prefix Metric
+ *
+ * TBD: MUST not be used for prefixes advertised as SRv6 locators.
+ *      This check has not been implemented yet.
+ *      (requirement is in RFC9350 Section 6.4
+ *      IS-IS Flexible Algorithm Definition Flags Sub-TLV)
+ *
+ * Look up the numbered flex-algo prefix metric associated
+ * with this prefix.
+ *
+ * This kind of flex-algo metric is advertised with various
+ * IP reachability TLVs for external routes.
+ *
+ * draft-ietf-lsr-flex-alg Section 8
+ * "IS-IS Flexible Algorithm Prefix Metric Sub-TLV"
+ */
+bool isis_flex_algo_prefix_metric(struct isis_subtlvs *subtlvs,
+				  uint32_t igp_metric, struct isis_area *area,
+				  uint8_t algo, uint32_t *metric)
+{
+	struct isis_flex_algo_prefix_metric *pm;
+	struct isis_router_cap_fad *fad;
+
+	fad = isis_flex_algo_elected_supported(algo, area);
+	if (!fad) {
+		zlog_debug(
+			"%s: no Flex-Algo Definition for the algorithm %u in area %s",
+			__func__, algo, area->area_tag);
+		return false;
+	}
+
+
+	/*
+	 * See extensive discussion in RFC9350
+	 * Section 13.1 Multi-area and Multi-domain Considerations
+	 *
+	 * If we ARE doing flex-algo, then use of the FAPM depends on the
+	 * flex-algo definition (FAD) M-flag.
+	 *
+	 * M-flag set: MUST use flex-algo prefix metric
+	 * M-flag clear: MUST use IGP metric
+	 */
+	if (!CHECK_FLAG(fad->fad.flags, FAD_FLAG_M)) {
+		/*
+		 * M-flag is not set, so use IGP metric
+		 */
+		*metric = igp_metric;
+		return true;
+	}
+
+	/*
+	 * FAD M-flag is set, so we MUST find the algo-specific prefix metric
+	 */
+
+	if (!subtlvs)
+		return false;
+
+	for (pm = (struct isis_flex_algo_prefix_metric *)
+			  subtlvs->flex_algo_prefix_metrics.head;
+	     pm; pm = pm->next) {
+		if (pm->algorithm == algo) {
+			*metric = pm->metric;
+			return true;
+		}
+	}
+	return false;
+}
+
+/*
+ * Extended IS reachability metric
+ *
+ * Given:
+ *
+ *	- the flex-algo number set in the spftree, and
+ *	- a given extended IS reachability TLV
+ *
+ * look up the algo-specific metric in the extended IS reachability subtlvs.
+ */
+bool isis_flex_algo_extended_is_metric(struct isis_spftree *spftree,
+				       uint32_t igp_metric,
+				       struct isis_ext_subtlvs *ies,
+				       uint32_t *metric)
+{
+	struct isis_router_cap_fad *fad;
+	struct isis_asla_subtlvs *asla;
+	uint32_t min_delay, te_metric;
+	struct listnode *node;
+
+	fad = isis_flex_algo_elected_supported(spftree->algorithm,
+					       spftree->area);
+	if (!fad) {
+		zlog_debug(
+			"%s: no Flex-Algo Definition for the algorithm %u in area %s",
+			__func__, spftree->algorithm, spftree->area->area_tag);
+		return false;
+	}
+
+	for (ALL_LIST_ELEMENTS_RO(ies->aslas, node, asla)) {
+		if (!CHECK_FLAG(asla->standard_apps, ISIS_SABM_FLAG_X))
+			continue;
+		break;
+	}
+
+	if (asla && CHECK_FLAG(asla->standard_apps, ISIS_SABM_FLAG_X)) {
```

Comment:
Here, you can't be sure that asla is set or not. First, in declaration asla is not initialized to NULL, second, the previous for loop could lead asla initialized or not depending if the loop finish at the end of the list of with the break.

Reply:
ALL_LIST_ELEMENTS_RO initializes asla to NULL and sets it to NULL at the end.
The code is correct but can be clarified

Reply:
done